api/src/main/java/io/druid/data/input/impl/prefetch/Fetcher.java

```diff
@@ -0,0 +1,322 @@
+  )
+  {
+    this.cacheManager = cacheManager;
+    this.objects = objects;
+    this.fetchExecutor = fetchExecutor;
+    this.temporaryDirectory = temporaryDirectory;
+    this.maxFetchCapacityBytes = maxFetchCapacityBytes;
+    this.prefetchTriggerBytes = prefetchTriggerBytes;
+    this.fetchTimeout = fetchTimeout;
+    this.maxFetchRetry = maxFetchRetry;
+    this.openObjectFunction = openObjectFunction;
+
+    this.prefetchEnabled = maxFetchCapacityBytes > 0;
+    this.numRemainingObjects = objects.size();
+
+    // (*) If cache is initialized, put all cached files to the queue.
+    this.fetchedFiles.addAll(cacheManager.getFiles());
+    this.nextFetchIndex = fetchedFiles.size();
+
+    if (prefetchEnabled) {
+      fetchIfNeeded(0L);
+    }
+  }
+
+  /**
+   * Submit a fetch task if remainingBytes is smaller than {@link #prefetchTriggerBytes}.
+   */
+  private void fetchIfNeeded(long remainingBytes)
+  {
+    if ((fetchFuture == null || fetchFuture.isDone())
+        && remainingBytes <= prefetchTriggerBytes) {
+      fetchFuture = fetchExecutor.submit(() -> {
+        fetch();
+        return null;
+      });
+    }
+  }
+
+  /**
+   * Fetch objects to a local disk up to {@link PrefetchableTextFilesFirehoseFactory#maxFetchCapacityBytes}.
+   * This method is not thread safe and must be called by a single thread.  Note that even
+   * {@link PrefetchableTextFilesFirehoseFactory#maxFetchCapacityBytes} is 0, at least 1 file is always fetched.
+   * This is for simplifying design, and should be improved when our client implementations for cloud storages
+   * like S3 support range scan.
+   *
+   * This method is called by {@link #fetchExecutor} if prefetch is enabled.  Otherwise, it is called by the same
+   * thread.
+   */
+  private void fetch() throws Exception
+  {
+    for (; nextFetchIndex < objects.size() && fetchedBytes.get() <= maxFetchCapacityBytes; nextFetchIndex++) {
+      final T object = objects.get(nextFetchIndex);
+      LOG.info("Fetching [%d]th object[%s], fetchedBytes[%d]", nextFetchIndex, object, fetchedBytes.get());
+      final File outFile = File.createTempFile(FETCH_FILE_PREFIX, null, temporaryDirectory);
+      fetchedBytes.addAndGet(download(object, outFile, 0));
+      fetchedFiles.put(new FetchedFile<>(object, outFile, getFileCloser(outFile, fetchedBytes)));
+    }
+  }
+
+  /**
+   * Downloads an object. It retries downloading {@link PrefetchableTextFilesFirehoseFactory#maxFetchRetry}
+   * times and throws an exception.
+   *
+   * @param object   an object to be downloaded
+   * @param outFile  a file which the object data is stored
+   * @param tryCount current retry count
+   *
+   * @return number of downloaded bytes
+   */
+  private long download(T object, File outFile, int tryCount) throws IOException
+  {
+    try (final InputStream is = openObjectFunction.open(object);
+         final OutputStream os = new FileOutputStream(outFile)) {
+      return IOUtils.copyLarge(is, os);
```

Comment:
Could you reuse buffer for all downloads in this Fetcher?

Reply:
Fixed.